### ISSUE TITLE
Question about using Ranger21 with Hugging Face Trainer

### ISSUE THREAD

**Issue by aliencaocao** [kozistr/pytorch_optimizer#167]
First of all, sorry if this issue is not very related to this repo, but I thought you being a professional at PyTorch optimizers, would know more on this than the rest.

I am not sure if you are familiar with Hugging Face Trainer https://huggingface.co/docs/transformers/v4.29.1/en/main_classes/trainer#transformers.Trainer , but they allow custom optimizer but also require a learning rate scheduler to be passed in, like
```
from transformers.optimization import Adafactor, AdafactorSchedule

optimizer = Adafactor(model.parameters(), scale_parameter=True, relative_step=True, warmup_init=True, lr=None)
lr_scheduler = AdafactorSchedule(optimizer)
trainer = Trainer(..., optimizers=(optimizer, lr_scheduler))
```

In this case, I can replace `optimizer` with `Ranger21`, but since `Ranger21` has built-in scheduling, how does this interact with the extra `lr_scheduler`? And what should I be passing in for that?

**Comment by kozistr** [comment]
You're right! `Ranger21` optimizer has a [built-in scheduler](https://github.com/kozistr/pytorch_optimizer/blob/main/pytorch_optimizer/optimizer/ranger21.py#L243) that only adjusts the learning rate at the start & end of the training (`k` < warm-up steps, `k` > warm-down steps). (You can check the details in README.md, `Linear learning rate warmup`, `Explore-exploit learning rate schedule`)

If you apply another lr scheduler like a `cosine annealing scheduler w/ warm-up`, the learning rate at the warmup stage works slightly differently (or same) (applied lr scheduler warm-up logic + built-in lr scheduler), and After the stage, only applied lr scheduler affects to the learning rate.

In short, I recommend using lr scheduler with Ranger21, and the built-in scheduler won't affect much I assume! (If you are worried about lr at the warm-up stage, use lr scheduler without a warmup or set `num_warm_up/down_iterations` parameter to 0)

I hope this could help and feel free to leave a comment if you have any questions or suggestions :)

thank you!

**Comment by aliencaocao** [comment]
Thanks for the answer. So to confirm my understanding, Ranger21's scheduler only modifies the start and end, and this library does it internally within the optimizer. Does this mean that for Hugging Face's side, if I pass in a constant LR schedule (meaning the LR stays at 1e-3 for example), Ranger21 internally will adjust the LR from this 1e-3 for the warm up and warm down, while the middle part will stay 1e-3?

Asking this because Hugging Face requires me to pass in a scheduler, and I do not want it to do anything to the Ranger21's built in one, to keep Ranger21 as factory-default as possible.

**Comment by kozistr** [comment]
> Thanks for the answer. So to confirm my understanding, Ranger21's scheduler only modifies the start and end, and this library does it internally within the optimizer. Does this mean that for Hugging Face's side, if I pass in a constant LR schedule (meaning the LR stays at 1e-3 for example), Ranger21 internally will adjust the LR from this 1e-3 for the warm up and warm down, while the middle part will stay 1e-3?
> 
> Asking this because Hugging Face requires me to pass in a scheduler, and I do not want it to do anything to the Ranger21's built in one, to keep Ranger21 as factory-default as possible.

`Ranger21 internally will adjust the LR from this 1e-3 for the warm up and warm down, while the middle part will stay 1e-3?` -> correct!

Maybe you can build a dummy learning rate scheduler and pass it to HF Trainer I guess!
```
lr_scheduler = LambdaLR(optimizer, lambda=lambda epoch: epoch)
```

**Comment by aliencaocao** [comment]
Yep thats what I have in mind. Thank you!
